gpflow/models/cglb.py

```diff
@@ -0,0 +1,158 @@
+        logsigma_sq = num_data * tf.math.log(sigma_sq)
+
+        # Correction term from Jensen's inequality
+        logtrace = num_data * tf.math.log(1 + trace / num_data)
+        return -output_dim * (logdet_b + 0.5 * logsigma_sq + 0.5 * logtrace)
+
+    def _quad_term(self, common) -> tf.Tensor:
+        x, y = self.data
+        err = y - self.mean_function(x)
+        num_data = to_default_float(tf.shape(err)[0])
+        sigma_sq = self.likelihood.variance
+        K = self.kernel.K(x) + sigma_sq * tf.eye(num_data, dtype=default_float())
+
+        A = common.A
+        LB = common.LB
+        preconditioner = NystromPreconditioner(A, LB, sigma_sq)
+        err_t = tf.transpose(err)
+
+        v_init = self._v
+        if not v_init.trainable:
+            v = cglb_conjugate_gradient(
+                K,
+                err_t,
+                v_init,
+                preconditioner,
+                self._max_cg_error,
+                self._max_cg_iters,
+                self._restart_cg_iters,
+            )
+        else:
+            v = v_init
+
+        Kv = v @ K
+        r = err_t - Kv
+        _, error_bound = preconditioner(r)
+        lb = tf.reduce_sum(v * (r + 0.5 * Kv))
+        ub = lb + 0.5 * error_bound
+
+        if not v_init.trainable:
+            v_init.assign(v)
+
+        return -ub
+
+
+class NystromPreconditioner:
+    """
+    Preconditioner of the form \math:`(Qff + σ²I)⁻¹`,
+    where \math:`A = σ⁻²L⁻¹Kᵤₓ` and \math:`B = AAᵀ + I`.
+    """
+
+    def __init__(self, A, LB, sigma_sq):
+        self.A = A
+        self.LB = LB
+        self.sigma_sq = sigma_sq
+
+    def __call__(self, v):
+        sigma_sq = self.sigma_sq
+        A = self.A
+        LB = self.LB
+
+        trans = tf.transpose
+        trisolve = tf.linalg.triangular_solve
+        matmul = tf.linalg.matmul
+        sum = tf.reduce_sum
+
+        v = trans(v)
+        Av = matmul(A, v)
+        LBinvAv = trisolve(LB, Av)
+        LBinvtLBinvAv = trisolve(trans(LB), LBinvAv, lower=False)
+
+        rv = v - matmul(A, LBinvtLBinvAv, transpose_a=True)
+        vtrv = sum(rv * v)
+        return trans(rv) / sigma_sq, vtrv / sigma_sq
+
+
+def cglb_conjugate_gradient(K, b, initial, preconditioner, max_error, max_steps, restart_cg_step):
+    """
+    :param A: [N, N] Matrix we want to backsolve from
+    :param b: [B, N] Vector we want to backsolve
+    :param P: Preconditioner (implements a mat_vec method)
+    :return: v: Approximate solution
+    """
+    CGState = namedtuple("CGState", ["i", "v", "r", "p", "rz"])
+
+    def stopping_criterion(state):
+        return (0.5 * state.rz > max_error) and (state.i < max_steps)
+
+    def cg_step(state):
+        Ap = state.p @ K
+        denom = tf.reduce_sum(state.p * Ap, axis=-1)
+        gamma = state.rz / denom
+        v = state.v + gamma * state.p
+        i = state.i + 1
+        r = tf.cond(
+            state.i % restart_cg_step == restart_cg_step - 1,
```

Comment:
Is restarting necessary? I think we ended up not using it in practice. It might be good to check and remove it if not.

Reply:
Indeed, the restarting is never happened in practice. But I'm a bit hesitant to remove it. We know it can be effective as a preventative measure.